Support custom Python classes in CUDAFuture

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56518 Add tests for CUDAFuture
* **#56516 Support custom Python classes in CUDAFuture**

One problem with CUDAFuture's extraction of DataPtrs from IValues is that it only supported Python objects that could be converted to "regular" IValues (e.g., lists/dicts/tuples of ints/strings/tensors/...). One notable exception are custom Python classes, which are in fact a very common data type transferred over RPC. The only solution we found for those is to use the Python pickler to extract the tensors contained in them.

We can't insert a Python dependency directly into CUDAFuture, so instead I'm proposing to use the same indirection technique used to support `getSubValues` on Python objects: define some methods on the abstract class `PyObjectHolder` (which can be used by CUDAFuture) but only implement them in the concrete subclass `ConcretePyObjectHolder` (which is only built when Python support is enabled).

I am a bit worried about the performance toll of this (pickling isn't exactly known to be cheap) but I think we should start by providing a functionally complete API. We already have ideas on how to make this faster if needed, for example by having users provide a custom DataPtr extractor tailored to their class via a decorator. (Or just use TorchScript).

Differential Revision: [D27887189](https://our.internmc.facebook.com/intern/diff/D27887189/)

lw committed:
Support custom Python classes in CUDAFuture

One problem with CUDAFuture's extraction of DataPtrs from IValues is that it only supported Python objects that could be converted to "regular" IValues (e.g., lists/dicts/tuples of ints/strings/tensors/...). One notable exception are custom Python classes, which are in fact a very common data type transferred over RPC. The only solution we found for those is to use the Python pickler to extract the tensors contained in them.

We can't insert a Python dependency directly into CUDAFuture, so instead I'm proposing to use the same indirection technique used to support `getSubValues` on Python objects: define some methods on the abstract class `PyObjectHolder` (which can be used by CUDAFuture) but only implement them in the concrete subclass `ConcretePyObjectHolder` (which is only built when Python support is enabled).

I am a bit worried about the performance toll of this (pickling isn't exactly known to be cheap) but I think we should start by providing a functionally complete API. We already have ideas on how to make this faster if needed, for example by having users provide a custom DataPtr extractor tailored to their class via a decorator. (Or just use TorchScript).

Differential Revision: [D27887189](https://our.internmc.facebook.com/intern/diff/D27887189/)

[ghstack-poisoned]

rohan-varma commented on aten/src/ATen/cuda/CUDAFuture.cpp:
```diff
@@ -27,9 +27,24 @@ namespace {
+    value.getSubValues(sub_values);
+  } catch (const c10::TypeError& err) {
+    // We might not be able to extract subvalues from Python objects (e.g., if
+    // they are custom classes) but we can still extract tensors.
```
Does this mean `ConcretePyObjectHodler::extractTensors` is on a best-effort basis and can't extract tensors that might be contained in nested classes?

lw replied:
It is "best effort" in the sense that while it widely enlarges the set of objects that are now supported, there are still objects that we won't be able to handle. These are however just the unpicklable objects, which hopefully are quite uncommon. Note that if you have an unpicklable object you're already unable to use it, for example, as an argument or a return value of a RPC call, hence I think it's reasonable to also say you cannot use it as a return value of a callback.

> can't extract tensors that might be contained in nested classes?

This should be supported, since pickle will just enter those nested classes and inspect them too. If you have a specific example you're worried about I'd be happy to add it to the tests in #56518.

rohan-varma replied:
Ah ok, sounds good, I just misunderstood the comment.

rohan-varma reviewed (commented):
Would be great to add tests for this change, especially for custom python classes in gradient compression/RPC

mrshenli reviewed (commented):
Would I be correct if I assume we need the dataPtrs because we need to call `recordStreams` instead of because we need to enforce execution orders? Because if we just need to enforce execution orders, we can record events on all devices in the `devices_` field?

lw commented:
> Would be great to add tests for this change, especially for custom python classes in gradient compression/RPC

I'm adding a test for it in #56518. Let me know if that seems enough for you or if there's more cases you'd like to see covered.

> Would I be correct if I assume we need the dataPtrs because we need to call recordStreams instead of because we need to enforce execution orders? Because if we just need to enforce execution orders, we can record events on all devices in the devices_ field?

Yes, indeed, the core reason we need to keep extracting DataPtrs is because we must record them with the CUDA caching allocator. If somehow we had a way to avoid that requirement (and I don't think we do), we could indeed avoid the extraction.

Note however that recording an event on _all_ devices, without checking which ones are really used, could have some performance downsides in some rare cases, due to us introducing "artificial" dependencies between kernels. As an example: imagine callback `A` launching two kernels `KA0` on cuda:0 and `KA1` on cuda:1, and returning those two tensors; and imagine a chained child callback `B` which only operates on cuda:0, ignoring the tensor on cuda:1, and enqueues a kernel `KB0`. Now, even if the result of `B` only uses cuda:0, anyone waiting for the result of `B` would also wait for the (unrelated) result of `KA1`. Such a condition is probably rare enough we can ignore it, but it's still something to consider.

lw committed:
Update on "Support custom Python classes in CUDAFuture"


One problem with CUDAFuture's extraction of DataPtrs from IValues is that it only supported Python objects that could be converted to "regular" IValues (e.g., lists/dicts/tuples of ints/strings/tensors/...). One notable exception are custom Python classes, which are in fact a very common data type transferred over RPC. The only solution we found for those is to use the Python pickler to extract the tensors contained in them.

We can't insert a Python dependency directly into CUDAFuture, so instead I'm proposing to use the same indirection technique used to support `getSubValues` on Python objects: define some methods on the abstract class `PyObjectHolder` (which can be used by CUDAFuture) but only implement them in the concrete subclass `ConcretePyObjectHolder` (which is only built when Python support is enabled).

I am a bit worried about the performance toll of this (pickling isn't exactly known to be cheap) but I think we should start by providing a functionally complete API. We already have ideas on how to make this faster if needed, for example by having users provide a custom DataPtr extractor tailored to their class via a decorator. (Or just use TorchScript).

Differential Revision: [D27887189](https://our.internmc.facebook.com/intern/diff/D27887189/)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM. Shall we get some micro-benchmark numbers to see the overhead of pickling?

lw committed:
Update on "Support custom Python classes in CUDAFuture"


One problem with CUDAFuture's extraction of DataPtrs from IValues is that it only supported Python objects that could be converted to "regular" IValues (e.g., lists/dicts/tuples of ints/strings/tensors/...). One notable exception are custom Python classes, which are in fact a very common data type transferred over RPC. The only solution we found for those is to use the Python pickler to extract the tensors contained in them.

We can't insert a Python dependency directly into CUDAFuture, so instead I'm proposing to use the same indirection technique used to support `getSubValues` on Python objects: define some methods on the abstract class `PyObjectHolder` (which can be used by CUDAFuture) but only implement them in the concrete subclass `ConcretePyObjectHolder` (which is only built when Python support is enabled).

I am a bit worried about the performance toll of this (pickling isn't exactly known to be cheap) but I think we should start by providing a functionally complete API. We already have ideas on how to make this faster if needed, for example by having users provide a custom DataPtr extractor tailored to their class via a decorator. (Or just use TorchScript).

Differential Revision: [D27887189](https://our.internmc.facebook.com/intern/diff/D27887189/)

[ghstack-poisoned]

lw commented:
> Shall we get some micro-benchmark numbers to see the overhead of pickling?

Good call. I used IPython to benchmark this as follows:
```
In [3]: t = torch.empty((100,), device="cuda:0")                                                                                                                                                                                                                                    

In [4]: def test_future(): 
   ...:     fut = Future(devices=["cuda:0"]) 
   ...:     fut.set_result([t]) 
   ...:     fut.value() 
   ...:                                                                                                                                                                                                                                                                             

In [5]: %timeit test_future()  
```

Before this PR, that code took 25.2us to run. On the original version of this PR (the one you reviewed yesterday) the performance was muuuch poorer: 19.4ms (1000x slower!!). I dug into that and it turned out the main culprit was the exception throwing/catching (i.e., we call getSubValues, which throws a TypeError, we catch it, and fall back to pickling). I replaced that by _always_ pickling in case of Python objects (without even trying to call getSubValues) and this helped dramatically: the code now took 60us. I played with it a bit more, and found out that defining the tensor extraction in Python (rather than in C++ with pybind11) helps a lot: the code now takes just 13.3us. It ended up being faster with pickling than it was before!! I still don't understand why using pybind11 rather than pure Python is slower, but I'll accept it.

lw committed:
Update on "Support custom Python classes in CUDAFuture"


One problem with CUDAFuture's extraction of DataPtrs from IValues is that it only supported Python objects that could be converted to "regular" IValues (e.g., lists/dicts/tuples of ints/strings/tensors/...). One notable exception are custom Python classes, which are in fact a very common data type transferred over RPC. The only solution we found for those is to use the Python pickler to extract the tensors contained in them.

We can't insert a Python dependency directly into CUDAFuture, so instead I'm proposing to use the same indirection technique used to support `getSubValues` on Python objects: define some methods on the abstract class `PyObjectHolder` (which can be used by CUDAFuture) but only implement them in the concrete subclass `ConcretePyObjectHolder` (which is only built when Python support is enabled).

I am a bit worried about the performance toll of this (pickling isn't exactly known to be cheap) but I think we should start by providing a functionally complete API. We already have ideas on how to make this faster if needed, for example by having users provide a custom DataPtr extractor tailored to their class via a decorator. (Or just use TorchScript).

Differential Revision: [D27887189](https://our.internmc.facebook.com/intern/diff/D27887189/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/jit/python/python_ivalue.h:
```diff
@@ -42,6 +42,19 @@ struct C10_EXPORT ConcretePyObjectHolder final : PyObjectHolder {
+    // We could implement this entirely in C++ via pybind11 but it turns out to
+    // be substantially slower. Namely, the total time taken by markCompleted on
+    // a CUDAFuture is 21.5us with this implementation, but goes up to 58.7us
+    // when using C++. The reason is unclear.
```
It might be due to more frequent C++ - Python context switches?

lw replied:
That could be one explanation. Another possibility could be that Python byte-compiles and somewhat optimizes the pure Python code? It sounds unlikely though...

I guess that something is going on here beyond what I'm currently understanding, and we might want to investigate, but I also feel we'd be going down a rabbit hole...

mrshenli commented on aten/src/ATen/cuda/CUDAFuture.cpp:
```diff
@@ -26,15 +26,26 @@ namespace {
+  // custom subclasses. Thus, instead, we extract the tensors through pickling.
+  if (value.isPyObject()) {
+    std::vector<at::Tensor> tensors =
+        value.toPyObjectHolder()->extractTensors();
```
hmm, could you please remind me how does the build target dependency work? I thought this file is in libtorch_cuda and the `extractTensors` is in `libtorch_python`. If that's the case, does it mean we need to make CUDA target depends on Python target?

lw replied:
So, I believe this works thanks to the fact that the `ConcretePyObjectHolder`, which is where the pickling actually happens, is in `libtorch_python` but its abstract base class `PyObjectHolder` is in `libtorch_cpp_cpu` (or whatever that's called). Thus CUDAFuture, which lives in `libtorch_cuda`, has access to the abstract base class, even though it doesn't have access to the concrete subclass. This is enough for the above code to build.

mrshenli reviewed (commented):
Great to know we got great perf here :)

My main question was on build target dependency.

lw commented on torch/csrc/jit/python/python_ivalue.h:
```diff
@@ -42,6 +42,19 @@ struct C10_EXPORT ConcretePyObjectHolder final : PyObjectHolder {
+    // be substantially slower. Namely, the total time taken by markCompleted on
+    // a CUDAFuture is 21.5us with this implementation, but goes up to 58.7us
+    // when using C++. The reason is unclear.
+    // FIXME There probably is a better package where this code could reside?
```
Since in another comment you mentioned dependencies between targets, I'm even more convinced that putting this Python helper function in the RPC module is a bad idea, because it basically means that JIT now has a dependency on RPC. The correct way to do this is to put this Python helper somewhere in a JIT module. Do you know what a good place would be?

mrshenli replied:
I am not sure if there is a better place to put this, but I agree with you, this is certainly better than having this in RPC, and I don't see any issue for the current version

lw committed:
Update on "Support custom Python classes in CUDAFuture"


One problem with CUDAFuture's extraction of DataPtrs from IValues is that it only supported Python objects that could be converted to "regular" IValues (e.g., lists/dicts/tuples of ints/strings/tensors/...). One notable exception are custom Python classes, which are in fact a very common data type transferred over RPC. The only solution we found for those is to use the Python pickler to extract the tensors contained in them.

We can't insert a Python dependency directly into CUDAFuture, so instead I'm proposing to use the same indirection technique used to support `getSubValues` on Python objects: define some methods on the abstract class `PyObjectHolder` (which can be used by CUDAFuture) but only implement them in the concrete subclass `ConcretePyObjectHolder` (which is only built when Python support is enabled).

I am a bit worried about the performance toll of this (pickling isn't exactly known to be cheap) but I think we should start by providing a functionally complete API. We already have ideas on how to make this faster if needed, for example by having users provide a custom DataPtr extractor tailored to their class via a decorator. (Or just use TorchScript).

Differential Revision: [D27887189](https://our.internmc.facebook.com/intern/diff/D27887189/)

[ghstack-poisoned]

mrshenli reviewed (changes requested):
Test failure does look relevant:

```
Apr 22 19:38:23   test_ddp_hook_parity_allreduce (__main__.TestDistBackendWithSpawn) ... SIGSEGV(11), PID: 37971, Thread 37971: 
Apr 22 19:38:23 frame #0: <unknown function> + 0x27d2f05 (0x7fa9b3acff05 in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so)
Apr 22 19:38:23 frame #1: <unknown function> + 0x27d3ac7 (0x7fa9b3ad0ac7 in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so)
Apr 22 19:38:23 frame #2: <unknown function> + 0x11390 (0x7fa9f76da390 in /lib/x86_64-linux-gnu/libpthread.so.0)
Apr 22 19:38:23 frame #3: <unknown function> + 0x168a52 (0x55844f831a52 in /opt/conda/bin/python)
Apr 22 19:38:23 frame #4: <unknown function> + 0x169398 (0x55844f832398 in /opt/conda/bin/python)
Apr 22 19:38:23 frame #5: <unknown function> + 0x3a008 (0x7fa9f7339008 in /lib/x86_64-linux-gnu/libc.so.6)
Apr 22 19:38:23 frame #6: <unknown function> + 0x3a055 (0x7fa9f7339055 in /lib/x86_64-linux-gnu/libc.so.6)
Apr 22 19:38:23 frame #7: <unknown function> + 0x20f2d9 (0x55844f8d82d9 in /opt/conda/bin/python)
Apr 22 19:38:23 frame #8: <unknown function> + 0x20f3b8 (0x55844f8d83b8 in /opt/conda/bin/python)
Apr 22 19:38:23 frame #9: PyErr_PrintEx + 0x32 (0x55844f8d8422 in /opt/conda/bin/python)
Apr 22 19:38:23 frame #10: PyRun_SimpleStringFlags + 0x66 (0x55844f8de706 in /opt/conda/bin/python)
Apr 22 19:38:23 frame #11: Py_Main + 0x423 (0x55844f8e24e3 in /opt/conda/bin/python)
Apr 22 19:38:23 frame #12: main + 0xee (0x55844f7acb0e in /opt/conda/bin/python)
Apr 22 19:38:23 frame #13: __libc_start_main + 0xf0 (0x7fa9f731f840 in /lib/x86_64-linux-gnu/libc.so.6)
Apr 22 19:38:23 frame #14: <unknown function> + 0x1c2d70 (0x55844f88bd70 in /opt/conda/bin/python)
```

lw committed:
Update on "Support custom Python classes in CUDAFuture"


One problem with CUDAFuture's extraction of DataPtrs from IValues is that it only supported Python objects that could be converted to "regular" IValues (e.g., lists/dicts/tuples of ints/strings/tensors/...). One notable exception are custom Python classes, which are in fact a very common data type transferred over RPC. The only solution we found for those is to use the Python pickler to extract the tensors contained in them.

We can't insert a Python dependency directly into CUDAFuture, so instead I'm proposing to use the same indirection technique used to support `getSubValues` on Python objects: define some methods on the abstract class `PyObjectHolder` (which can be used by CUDAFuture) but only implement them in the concrete subclass `ConcretePyObjectHolder` (which is only built when Python support is enabled).

I am a bit worried about the performance toll of this (pickling isn't exactly known to be cheap) but I think we should start by providing a functionally complete API. We already have ideas on how to make this faster if needed, for example by having users provide a custom DataPtr extractor tailored to their class via a decorator. (Or just use TorchScript).

Differential Revision: [D27887189](https://our.internmc.facebook.com/intern/diff/D27887189/)

[ghstack-poisoned]

lw committed:
Update on "Support custom Python classes in CUDAFuture"


One problem with CUDAFuture's extraction of DataPtrs from IValues is that it only supported Python objects that could be converted to "regular" IValues (e.g., lists/dicts/tuples of ints/strings/tensors/...). One notable exception are custom Python classes, which are in fact a very common data type transferred over RPC. The only solution we found for those is to use the Python pickler to extract the tensors contained in them.

We can't insert a Python dependency directly into CUDAFuture, so instead I'm proposing to use the same indirection technique used to support `getSubValues` on Python objects: define some methods on the abstract class `PyObjectHolder` (which can be used by CUDAFuture) but only implement them in the concrete subclass `ConcretePyObjectHolder` (which is only built when Python support is enabled).

I am a bit worried about the performance toll of this (pickling isn't exactly known to be cheap) but I think we should start by providing a functionally complete API. We already have ideas on how to make this faster if needed, for example by having users provide a custom DataPtr extractor tailored to their class via a decorator. (Or just use TorchScript).

Differential Revision: [D27887189](https://our.internmc.facebook.com/intern/diff/D27887189/)

[ghstack-poisoned]
